test/boost/memtable_test.cc

```diff
@@ -72,6 +73,28 @@ std::vector<mutation> make_ring(schema_ptr s, int n_mutations) {
     return ring;
 }
 
+static mutation_source make_memtable_reverse_mutation_source(lw_shared_ptr<memtable>&& mt) {
+    return mutation_source([mt = std::move(mt)] (schema_ptr s,
+                                      reader_permit permit,
+                                      const dht::partition_range& pr,
+                                      const query::partition_slice& slice,
+                                      const io_priority_class& pc,
+                                      tracing::trace_state_ptr tr,
+                                      streamed_mutation::forwarding fwd,
+                                      mutation_reader::forwarding mr_fwd) {
+        // TODO: now it is assumed that the memtable takes slices in half-reversed format (i.e. [4, 6], [2, 3]).
+        // This is the format needed by most of the memtable code. In future, it is expected, that the queries 
+        // will receive native-reversed format ([6, 4], [3, 2]). Change the following line when this change happens.
```

Comment:
This situation where some readers expect one format of ranges (half reversed, native reversed) and some another format is very bad. It's hard to track this. It makes higher layers need to be aware of reader implementations. It makes conversion harder, since I need to selectively switch higher layers to a different format. It means you cannot just replace reader implementation at higher layers. The interface should be uniform in that regard (it's part of the `mutation_source` abstraction). We should switch all readers to expect one format, and those who want a different format should convert on their own.

\cc @kbr- @denesb

Reply:
I agree and this is what I had in my series originally, but @kbr- asked me to drop this to make intermediary patches simpler, however it seems this has a cost in safety, which I'm not sure is worth it.

Reply:
In addition to safety, I don't know which callers to change when I want to transition to a new format. Since the reader is behind an abstract interface (mutation_source), there can be non-obvious users. It seems simpler to me to do a sweeping change at the interface level, leaving conversions in the implementations, and then you drop the conversion when you're working on a particular implementation.

Reply:
AFAIK right now, and after my patch, and after @enedil's patch, all readers still expect the half reversed format so nothing has changed. Which readers expect the native format?

Reply:
Ok one thing has changed: they now expect the reversed schema.

So in the end all readers expect:
- the reversed schema,
- and a half-reversed slice.

Reply:
It will be simple to add a single commit at the end - when both my and @enedil's changes are integrated - that modifies all readers from expecting half-reversed slices to expecting fully/native-reversed slices.

Reply:
> It makes higher layers need to be aware of reader implementations.

Readers should specify at the interface level which format they expect.
For example, consider this comment from my patchset (in `sstable` class):
```
    // Precondition: if the slice is reversed, the schema must be reversed as well.
    // Reversed slices must be provided in the 'half-reversed' format (the order of ranges
    // being reversed, but the ranges themselves are not).
    //
    // If the slice is reversed, each partition returned by the reader will be reversed,
    // i.e. the fragments will be returned according to the order of the reversed schema,
    // and the reader does not support fast forwarding.
    flat_mutation_reader make_reader_v1(
```

No need to look into the implementation.

Reply:
> > It makes higher layers need to be aware of reader implementations.
> 
> Readers should specify at the interface level which format they expect.
> For example, consider this comment from my patchset (in `sstable` class):
> 
> ```
>     // Precondition: if the slice is reversed, the schema must be reversed as well.
>     // Reversed slices must be provided in the 'half-reversed' format (the order of ranges
>     // being reversed, but the ranges themselves are not).
>     //
>     // If the slice is reversed, each partition returned by the reader will be reversed,
>     // i.e. the fragments will be returned according to the order of the reversed schema,
>     // and the reader does not support fast forwarding.
>     flat_mutation_reader make_reader_v1(
> ```
> 
> No need to look into the implementation.

You can do this legally, but it makes things harder because now a partition_slice which I pass to one reader may not fit to another reader. I have to at least consult the interface doc for each reader and apply conversions if necessary. It's a burden and a potential for bugs. It's better if I don't have to do this, and just pass the slice along trusting that all readers will interpret it in the same way.

We also must come up with a uniform format for the mutation_source interface, since it erases the information about the implementation. You have sstable::as_mutation_source() for example, and that mutation source is then passed to contexts which don't know which impl it contains. It's a risk to have make_reader() differ from mutation_source, since it's easy to get as_mutation_source() wrong just by not paying attention to it, it usually just calls make_reader(). make_reader() should actually be an implementation of virtual mutation_source base class.

It's true that you can have a patch which switches readers to accept a native format afterwards, but I think it would be safer and less burden-some to have such a patch beforehand.

Reply:
I agree with everything you said except the last paragraph.

The coordinator sends us a half-reversed slice and everyone expects the half-reversed slice prior to our changes.
At each step of our changes this does not change: everyone keeps expecting half-reversed slice. Low-level source readers (sstable and memtable) can convert the half-reversed slice to a native-reversed slice when they need it internally. So we introduce 2 conversions that are internal to these sources, each the moment where they need it (i.e. where they switch to the new reverse mode), without changing the interfaces.

Regarding the first two paragraphs - I agree the mutation_source interfaces should be uniform, and they are - we expect half-reversed slices everywhere.

Introducing a conversion from half-reversed to native-reversed at the top of the replica query stack would require more work to avoid the exact problem you describe: the same commit which introduces the conversion would have to simultaneously modify the source readers at the bottom of the stack so they convert the native reverse slice to a half reverse slice and store it internally. Then we have to gradually remove these bottom conversions when the readers implement the new reverse mode. It seems that we need to perform a similar amount of work but at different points in time. I prefer the lazy approach and do the work as late as possible since that's when it's most motivated and it delays the API change to the very end. And I think the first iteration sent to customers should not contain an API change as this minimizes the chance of introducing bugs - API changes are bug-prone.

@denesb 
> I agree and this is what I had in my series originally, but @kbr- asked me to drop this to make intermediary patches simpler, however it seems this has a cost in safety, which I'm not sure is worth it.

IIRC you added a conversion on top of the replica reader stack and then another conversion back to half-reversed in the middle of the stack (in querier), basically introducing the exact problem that @tgrabiec is describing: we had some readers at the upper part of the stack expecting native reversed format and some readers at the lower part of the stack expecting half reversed format. This made my intermediary patches hard because I had to push the point of API change (slice conversion) lower and lower, and storing a slice is burdensome.

The current approach assumes half-reversed everywhere.